### PR TITLE
tree2: Allow inserting content into editable tree with cursors

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Editable tree now excepts new content in the form of cursors.

Alos included are some testing cleanups to avoid depending on SharedTree for testing editable tree, and some improved coverage on the editing methods of fields which were previously entirely untested.

## Breaking Changes

Code that was violating the type system and providing cursors already for list insertion methods happened to work, but now requires a single field cursor instead of an array of node cursors (and no longer violates the typing).

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts [hunk 1]
```diff
@@ -513,21 +513,33 @@ export type AssignableFieldKinds = typeof FieldKinds.optional | typeof FieldKind
...
-export type FlexibleNodeContent<TTypes extends AllowedTypes> = SchemaAware.AllowedTypesToTypedTrees<
-	SchemaAware.ApiMode.Flexible,
-	TTypes
->;
+export type FlexibleNodeContent<TTypes extends AllowedTypes> =
+	| SchemaAware.AllowedTypesToTypedTrees<SchemaAware.ApiMode.Flexible, TTypes>
+	| ITreeCursorSynchronous;
+
+/**
+ * Strongly typed tree literals for inserting a subsequence of nodes.
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:535]: nit: "subsequence" can be interpreted as "a range within a sequence" but also "a sequence element of a sequence of sequences". My brain gravitated toward the second definition without any other context at first, FWIW. You may want to clarify what a sub-sequence is, here.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/utils.ts [hunk 2]
```diff
@@ -38,31 +46,44 @@ export function getReadonlyContext(forest: IEditableForest, schema: TreeSchema):
 /**
  * Creates a context and its backing forest from the provided `content`.
  *
+ * For creating mutable views use {@link viewWithContent}, but prefer this when possible as its has fewer dependencies and simpler setup.
```
Comment (noencke) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/utils.ts:49]:
```suggestion
 * For creating mutable views use {@link viewWithContent}, but prefer this when possible as it has fewer dependencies and simpler setup.
```

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/utils.ts [hunk 3]
```diff
@@ -76,17 +97,26 @@ export function makeSchema<const TSchema extends ImplicitFieldSchema>(
 	return builder.intoSchema(root);
 }
 
+/**
+ * @deprecated Writing a normal `it` test. Doing so allows:
```
Comment (noencke) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/utils.ts:101]:
```suggestion
 * @deprecated Write a normal `it` test. Doing so allows:
```

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts [hunk 4]
```diff
@@ -103,7 +105,7 @@ function applySequenceFieldEdit(
 			const parent = navigateToNode(tree, change.parent);
 			assert(parent?.is(fuzzNode), "Defined down-path should point to a valid parent");
 			const field = parent.boxedSequenceChildren;
-			field.insertAt(change.index, [singleTextCursor(change.value) as any]);
+			field.insertAt(change.index, fieldCursorFromJsonableTrees([change.value]));
```
Comment (noencke) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts:108]: Should `fieldCursorFromJsonableTrees` be exposed outside of the chunked-forest area? Seems like a generally useful function now, like `singleTextCursor`.

**Review by noencke — APPROVED** [review]
Thanks for all the cleanup along the way!

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts (hunk 4 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts:108]: yes at some point, but probably not that implementation which has some odd side effects on the path. I have added some todos about this to treeCursor utils.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts (hunk 1 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:535]: Added some clarification.

Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts:535]: Thanks!

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts [hunk 5]
```diff
@@ -585,32 +582,32 @@ function buildStructClass<TSchema extends ObjectNodeSchema>(
...
 					this: CustomStruct,
-					newContent: ContextuallyTypedNodeData,
+					newContent: FlexibleNodeContent<AllowedTypes> | undefined,
 				): void {
 					const field = getBoxedField(
 						this,
 						key,
 						fieldSchema,
-					) as RequiredField<AllowedTypes>;
+					) as OptionalField<AllowedTypes>;
```
Comment (noencke) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/lazyTree.ts:596]: Lol :( nice catch

**Review by noencke — APPROVED** [review]

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts (hunk 4 above)
Comment (noencke) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts:108]: Sounds good, thanks.

**Pull request merged**